### PR TITLE
mapblock_mesh_generate_special improvements

### PR DESCRIPTION
**Pull request by gregorycu**
Remove uses of std::map
Make sure we use ContentFeatures reference not value

On my machine, Windows x64:
This speeds up mapblock_mesh_generate_special by 35%.
This function used to represent 24% of UpdateThread's CPU work
After these changes, it's down to 14%.
In real terms, the average call time was 1.1ms. Now it's 0.65ms.

### CODE REVIEW

**Comment by Fixer-007** [comment]
Compiler warning, it compiles successfully though. 
```
[ 35%] Building CXX object src/CMakeFiles/minetest.dir/content_mapblock.cpp.obj
/media/sf_Exchange/minetest/build/minetest/src/content_mapblock.cpp: In function ‘void mapblock_mesh_generate_special(MeshMakeData*, MeshCollector&)’:
/media/sf_Exchange/minetest/build/minetest/src/content_mapblock.cpp:489:12: warning: variable ‘neighbordir’ set but not used [-Wunused-but-set-variable]
      v3s16 neighbordir = cornerdir - halfdirs[j];
            ^
```

**Comment by Fixer-007** [comment]
Flowing liquids are bugged with this PR:
2016-09-24 01_50_05-minetest 0 4 14-cd2c8ad-cd2c8ad-dirty opengl 4 5 13416

**Comment by gregorycu** [comment]
I have update the PR. The warning and the bug were the same thing.

**Comment by Fixer-007** [comment]
obviously :)

**Comment by est31** [comment]
:+1: if the style issues are fixed (like the one I pointed out, but its present in other places of the PR as well).

File: src/content_mapblock.cpp [hunk 1]
```diff
@@ -172,6 +172,13 @@ static inline void getNeighborConnectingFace(v3s16 p, INodeDefManager *nodedef,
 		*neighbors |= v;
 }
 
+namespace
```
Comment (est31) [inline src/content_mapblock.cpp]: Please make the function static instead of using unnamed namespaces. It has the same outcome, but is more used in the code, and less "C++y".

Comment (gregorycu) [inline src/content_mapblock.cpp]: What's wrong with being idiomatic? Consistency is important, I suppose. (Although static does only work with functions, not classes and types etc.) Will make change, but don't understand the rationale.

File: src/content_mapblock.cpp [hunk 2]
```diff
@@ -1506,8 +1523,8 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
 						continue;
 					MapNode n_xy = data->m_vmanip.getNodeNoEx(blockpos_nodes + v3s16(x + xz, y + y0, z));
 					MapNode n_zy = data->m_vmanip.getNodeNoEx(blockpos_nodes + v3s16(x, y + y0, z + xz));
-					ContentFeatures def_xy = nodedef->get(n_xy);
-					ContentFeatures def_zy = nodedef->get(n_zy);
+					const ContentFeatures& def_xy = nodedef->get(n_xy);
```
Comment (gregorycu) [inline src/content_mapblock.cpp]: Will do.

File: src/content_mapblock.cpp [hunk 3]
```diff
@@ -472,10 +484,11 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
 				for(u32 j=0; j<4; j++)
 				{
 					v3s16 neighbordir = cornerdir - halfdirs[j];
-					content_t content = neighbor_contents[neighbordir];
+
+					NeighbourData &neighbour_data = neighbour_datas[NeightbourToIndex(neighbordir)];
+					content_t& content = neighbour_data.content;
```
Comment (est31) [inline src/content_mapblock.cpp]: missed this one

**Comment by est31** [comment]
:+1:

File: src/content_mapblock.cpp [hunk 4]
```diff
@@ -401,9 +405,15 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
...
-			std::map<v3s16, content_t> neighbor_contents;
-			std::map<v3s16, u8> neighbor_flags;
+
+			struct NeighbourData
+			{
+				f32 level;
+				content_t content;
+				u8 flags;
+			};
+			NeighbourData neighbour_datas[27];
```
Comment (Zeno-) [inline src/content_mapblock.cpp]: pedantic I know, but "neighbour_datas"? Data is, technically, already plural. The word "datas" just sounds... wrong. The aggregate object name is different to the struct name anyway, so what's wrong with "neighbour_data"?

File: src/content_mapblock.cpp [hunk 5]
```diff
@@ -526,15 +539,16 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
 			{
 				v3s16 dir = side_dirs[i];
 
+				NeighbourData& neighbour_data = neighbour_datas[NeightbourToIndex(dir)];
```
Comment (Zeno-) [inline src/content_mapblock.cpp]: NeightbourToIndex ... typo here (function should be called NeighbourToIndex() I believe)

File: src/content_mapblock.cpp [hunk 6]
```diff
@@ -172,6 +172,10 @@ static inline void getNeighborConnectingFace(v3s16 p, INodeDefManager *nodedef,
 		*neighbors |= v;
 }
 
+static int NeightbourToIndex(const v3s16 &pos) {
```
Comment (Zeno-) [inline src/content_mapblock.cpp]: See comment below regarding function name typo

**Comment by Zeno-** [comment]
:-1: until the typos are fixed

**Comment by gregorycu** [comment]
I have fixed the "NeightbourToIndex" typo.

File: src/content_mapblock.cpp [hunk 7]
```diff
@@ -172,6 +172,10 @@ static inline void getNeighborConnectingFace(v3s16 p, INodeDefManager *nodedef,
 		*neighbors |= v;
 }
 
+static int NeighbourToIndex(const v3s16 &pos) {
```
Comment (Zeno-) [inline src/content_mapblock.cpp]: Closing brace should be on next line.

Should this function be inline?

File: src/content_mapblock.cpp [hunk 8]
```diff
@@ -172,6 +172,10 @@ static inline void getNeighborConnectingFace(v3s16 p, INodeDefManager *nodedef,
...
+	return (9 * pos.X + 3 * pos.Y + pos.Z) + 13;
```
Comment (Zeno-) [inline src/content_mapblock.cpp]: What is 13?

Comment (Zeno-) [inline src/content_mapblock.cpp]: And, why would 13 (whatever it is) be added to (x=0,y=0,z=0)?

Comment (Zeno-) [inline src/content_mapblock.cpp]: This line makes no sense. If it's correct a comment is needed for sure

Comment (HybridDog) [inline src/content_mapblock.cpp]: ps = {-1; 0; 1}
pos.X ∈ ps; pos.Y ∈ ps; pos.Y ∈ ps
return_value ∈ {0; 1; …; 26}

Comment (HybridDog) [inline src/content_mapblock.cpp]: pos.X ∈ {-1; 0; 1}
pos.Y ∈ {-1; 0; 1}
pos.Z ∈ {-1; 0; 1}
NeighbourToIndex(pos) ∈ {0; 1; …; 26}

Comment (HybridDog) [inline src/content_mapblock.cpp]: 13 is used to make the return value >= 0

pos.X ∈ {-1; 0; 1}
pos.Y ∈ {-1; 0; 1}
pos.Z ∈ {-1; 0; 1}
NeighbourToIndex(pos) ∈ {0; 1; …; 26}

File: src/content_mapblock.cpp (hunk 7 above)
Comment (gregorycu) [inline src/content_mapblock.cpp]: It may be. Up to the compiler/linker.

File: src/content_mapblock.cpp (hunk 8 above)
Comment (Zeno-) [inline src/content_mapblock.cpp]: HyrbidDog. Correct. And those are undocumented preconditions on what x, y, and z can be.

File: src/content_mapblock.cpp (hunk 7 above)
Comment (Zeno-) [inline src/content_mapblock.cpp]: Should there be a suggestion to the compiler/linker that this is better inlined?

File: src/content_mapblock.cpp (hunk 8 above)
Comment (gregorycu) [inline src/content_mapblock.cpp]: I've added comments to better clarify how this function should be used. I've also added a comment to make it a lambda when we move to C++11, to remind us to move it into a smaller scope so it is less likely to be used outside the function. I'm not too sure about the bitshift to be honest. (In the sense I don't know - not that I disagree).

File: src/content_mapblock.cpp [hunk 9]
```diff
@@ -472,10 +488,11 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
...
 				{
 					v3s16 neighbordir = cornerdir - halfdirs[j];
-					content_t content = neighbor_contents[neighbordir];
+
+					NeighbourData &neighbour_data = neighbour_datas[NeighbourToIndex(neighbordir)];
+					content_t& content = neighbour_data.content;
 					// If top is liquid, draw starting from top of node
-					if(neighbor_flags[neighbordir] &
-							neighborflag_top_is_same_liquid)
+					if (neighbour_data.flags & neighborflag_top_is_same_liquid)
```
Comment (rubenwardy) [inline src/content_mapblock.cpp]: Sorry for nit picks, but two style errors:     `content_t &content` and { on same line as if

File: src/content_mapblock.cpp [hunk 10]
```diff
@@ -526,15 +543,16 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
...
 				v3s16 dir = side_dirs[i];
 
+				NeighbourData& neighbour_data = neighbour_datas[NeighbourToIndex(dir)];
 				/*
 					If our topside is liquid and neighbor's topside
 					is liquid, don't draw side face
 				*/
 				if(top_is_same_liquid &&
-						neighbor_flags[dir] & neighborflag_top_is_same_liquid)
+					neighbour_data.flags & neighborflag_top_is_same_liquid)
```
Comment (rubenwardy) [inline src/content_mapblock.cpp]: double intend on if logic overflow, space after if

File: src/content_mapblock.cpp (hunk 8 above)
Comment (kwolekr) [inline src/content_mapblock.cpp]: The bitshift method might be more intuitive, but much less efficient overall due to the index gaps causing more of a space requirement.  Multiplication by constant expressions are optimized by the compiler when it's possible to, so that argument is moot.  I'd rather not have something that burdensome instead because it "might possibly" aid with understanding.

I think a lot can be illustrated by demonstrating in comments where the form of this equation came from, e.g. "// This expression is a simplification of (3*3*1) * (pos.Z+1) + (3*1) * (pos.Y+1) + (1) * (pos.X+1)".

File: src/content_mapblock.cpp (hunk 7 above)
Comment (kwolekr) [inline src/content_mapblock.cpp]: Marking this function as inline probably won't have any effect on the compiler's decision.  What it will do is signal the author's intent on the intended role of this function.  So, I agree with Zeno` that maybe it should be marked inline. 
Also I think this function should be named "MeshGenerateSpecialNeighbourToIndex" to indicate its relation to that function too.

File: src/content_mapblock.cpp (hunk 4 above)
Comment (kwolekr) [inline src/content_mapblock.cpp]: Agreed, "datas" is ungrammatical.  "datum" wouldn't work either since it's a struct with three distinct datums that come together to form data, so it'd be misleading.  The best thing to do here would be to just rename the variable to "neighbour_data" IMHO.

File: src/content_mapblock.cpp [hunk 11]
```diff
@@ -401,9 +409,15 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
 
 			// Neighbor liquid levels (key = relative position)
 			// Includes current node
-			std::map<v3s16, f32> neighbor_levels;
-			std::map<v3s16, content_t> neighbor_contents;
-			std::map<v3s16, u8> neighbor_flags;
+
+			struct NeighbourData
+			{
```
Comment (kwolekr) [inline src/content_mapblock.cpp]: This should be on the same line.

File: src/content_mapblock.cpp [hunk 12]
```diff
@@ -472,10 +488,11 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
 				for(u32 j=0; j<4; j++)
 				{
 					v3s16 neighbordir = cornerdir - halfdirs[j];
-					content_t content = neighbor_contents[neighbordir];
+
+					NeighbourData &neighbour_data = neighbour_datas[NeighbourToIndex(neighbordir)];
+					content_t& content = neighbour_data.content;
```
Comment (kwolekr) [inline src/content_mapblock.cpp]: Why is this content_t a reference here?

Are you modifying it somewhere below?

**Comment by kwolekr** [comment]
Overall I really like what this PR does and how it goes about it, but there are some smaller nitpicky style issues left that should be addressed before merge.

File: src/content_mapblock.cpp (hunk 4 above)
Comment (gregorycu) [inline src/content_mapblock.cpp]: Except that would clash with another variable called neighbour_data. I would have happily used it otherwise. I'm not sure there are a good set of two names. My mental parsing of neighbour_datas is (neighbour_data)s. Do you have suggestions for both.

File: src/content_mapblock.cpp (hunk 12 above)
Comment (gregorycu) [inline src/content_mapblock.cpp]: No, it's probably an oversight. Will remove.

File: src/content_mapblock.cpp (hunk 9 above)
Comment (gregorycu) [inline src/content_mapblock.cpp]: Will fix.

File: src/content_mapblock.cpp (hunk 10 above)
Comment (gregorycu) [inline src/content_mapblock.cpp]: Sorry, I don't understand what you mean. Will fix the space.

File: src/content_mapblock.cpp (hunk 11 above)
Comment (gregorycu) [inline src/content_mapblock.cpp]: Will fix.

File: src/content_mapblock.cpp (hunk 10 above)
Comment (kwolekr) [inline src/content_mapblock.cpp]: He means that if the if statement logic goes onto the next line, it needs an extra indentation space, so you can easily disambiguate a continuation of the if statement logic from the if statement body.

File: src/content_mapblock.cpp (hunk 4 above)
Comment (kwolekr) [inline src/content_mapblock.cpp]: Ah, I see what you mean now.  What about something like neighbour_data_matrix?

Comment (gregorycu) [inline src/content_mapblock.cpp]: That's fantastic. Will change to that.

File: src/content_mapblock.cpp (hunk 1 above)
Comment (kwolekr) [inline src/content_mapblock.cpp]: Well because the anonymous namespace isn't used anywhere else.  If the rest of the code does something one way it's just best to go with the flow.

If you can make a solid argument why we should all dump static for anonymous namespace, then by all means make a PR for it that changes it everywhere.

File: src/content_mapblock.cpp [hunk 13]
```diff
@@ -172,6 +172,14 @@ static inline void getNeighborConnectingFace(v3s16 p, INodeDefManager *nodedef,
 		*neighbors |= v;
 }
 
+// For use in mapblock_mesh_generate_special
+// X,Y,Z of position must be -1,0,1
+// (If you don't know what this function does, you shouldn't be using it).
```
Comment (Zeno-) [inline src/content_mapblock.cpp]: That last line "(If you don't know what this function does, you shouldn't be using it)" is not appropriate. It's not about knowing what the function does, it's about documentation for future developers.

File: src/content_mapblock.cpp (hunk 8 above)
Comment (HybridDog) [inline src/content_mapblock.cpp]: You're right.
The function resembles the VoxelArea:index function.

File: src/content_mapblock.cpp [hunk 14]
```diff
@@ -172,6 +172,14 @@ static inline void getNeighborConnectingFace(v3s16 p, INodeDefManager *nodedef,
...
+static int NeighbourToIndex(const v3s16 &pos)
```
Comment (nerzhul) [inline src/content_mapblock.cpp]: Inline this function please

**Comment by paramat** [comment]
Author is unable to work on this for a while, i might attempt to finish this off.

File: src/content_mapblock.cpp (hunk 10 above)
Comment (paramat) [inline src/content_mapblock.cpp]: 'double indent' :]

File: src/content_mapblock.cpp (hunk 9 above)
Comment (paramat) [inline src/content_mapblock.cpp]: To be fair the whole file is full of curly brackets on wrong lines, missing spaces, maybe style cleanup can be another commit?

**Comment by paramat** [comment]
PR finished off in #4567 
@gregorycu it's annoying when someone disappears without finishing a PR, it's a difficult job to reproduce the diff and finish stuff when this happens ;]

**Comment by gregorycu** [comment]
@paramat Yeah. I know it sucks.

**Comment by paramat** [comment]
Anyway it's good to have a break, and probably essential for sanity, will miss you.

**Comment by paramat** [comment]
#4567 is merged and replaces this.
Thanks for the original commit.
